In [ ]:
#Menampilkan Drive
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


#1.Berdasarkan dataset Restaurant_Reviews.tsv yang sudah dibahas sebelumnya (ada di WAG Cherno). Buatlah model Bag of Words dengan semua model klasifikasi yang ada (termasuk Naive Bayes yang sudah dikerjakan), dan carilah model dengan tingkat akurasi paling tinggi.

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score


In [ ]:
df = pd.read_csv('/content/gdrive/MyDrive/Python TIA/ml-1m/Restaurant_Reviews.tsv', delimiter='\t', quoting=3)

In [ ]:
# Memisahkan fitur (review) dan label (liked)
X = df['Review'].values
y = df['Liked'].values


In [ ]:
df

Review  Liked
0                             Wow... Loved this place.      1
1                                   Crust is not good.      0
2            Not tasty and the texture was just nasty.      0
3    Stopped by during the late May bank holiday of...      1
4    The selection on the menu was great and so wer...      1
..                                                 ...    ...
995  I think food should have flavor and texture an...      0
996                           Appetite instantly gone.      0
997  Overall I was not impressed and would not go b...      0
998  The whole experience was underwhelming, and I ...      0
999  Then, as if I hadn't wasted enough of my life ...      0

[1000 rows x 2 columns]

In [ ]:
# Inisialisasi
vectorizer = CountVectorizer()

In [ ]:
# Menerapkan transformasi CountVectorizer pada data teks
X = vectorizer.fit_transform(X)


In [ ]:
# Membagi data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
# Inisialisasi model klasifikasi
naive_bayes = MultinomialNB()
logistic_regression = LogisticRegression(max_iter=1000)
svm = SVC()


In [ ]:
# Melatih dan menguji model Naive Bayes
naive_bayes.fit(X_train, y_train)
nb_predictions = naive_bayes.predict(X_test)
nb_accuracy = accuracy_score(y_test, nb_predictions)
print("Naive Bayes accuracy:", nb_accuracy)


Naive Bayes accuracy: 0.795


In [ ]:
# Melatih dan menguji model Logistic Regression
logistic_regression.fit(X_train, y_train)
lr_predictions = logistic_regression.predict(X_test)
lr_accuracy = accuracy_score(y_test, lr_predictions)
print("Logistic Regression accuracy:", lr_accuracy)

Logistic Regression accuracy: 0.81


In [ ]:
# Melatih dan menguji model SVM
svm.fit(X_train, y_train)
svm_predictions = svm.predict(X_test)
svm_accuracy = accuracy_score(y_test, svm_predictions)
print("SVM accuracy:", svm_accuracy)

SVM accuracy: 0.73


#Naive Bayes accuracy: 0.795
#Logistic Regression accuracy: 0.81
#SVM accuracy: 0.73
#Logistik Regression paling tinggi


#2.Coba tingkatkan lagi akurasinya dengan menyesuaikan daftar stopwords yang digunakan (seperti me-remove “not” dari daftar stopwords). Berapa tingkat akurasi tertinggi yang bisa dicapai?

In [ ]:
import nltk
from nltk.corpus import stopwords

In [ ]:
df = pd.read_csv('/content/gdrive/MyDrive/Python TIA/ml-1m/Restaurant_Reviews.tsv', delimiter='\t', quoting=3)

In [ ]:
# Memisahkan fitur (review) dan label (liked)
X = df['Review'].values
y = df['Liked'].values

In [ ]:
# Inisialisasi stopwords dalam bahasa Inggris
nltk.download('stopwords')
stopwords = set(stopwords.words('english'))
stopwords.remove('not')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [ ]:
# Mengubah teks menjadi lowercase dan menghapus stopwords
def preprocess_text(text):
    words = text.lower().split()
    words = [word for word in words if word not in stopwords]
    return ' '.join(words)

In [ ]:
# Melakukan preprocessing pada data teks
X = [preprocess_text(review) for review in X]

In [ ]:
# Inisialisasi
vectorizer = CountVectorizer()


In [ ]:
# Menerapkan transformasi
X = vectorizer.fit_transform(X)

In [ ]:
# Membagi data menjadi set pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [ ]:
# Inisialisasi model klasifikasi
naive_bayes = MultinomialNB()
logistic_regression = LogisticRegression(max_iter=1000)
svm = SVC()


In [ ]:
# Melatih dan menguji model Naive Bayes
naive_bayes.fit(X_train, y_train)
nb_predictions = naive_bayes.predict(X_test)
nb_accuracy = accuracy_score(y_test, nb_predictions)
print("Naive Bayes accuracy:", nb_accuracy)

Naive Bayes accuracy: 0.785


In [ ]:
# Melatih dan menguji model Logistic Regression
logistic_regression.fit(X_train, y_train)
lr_predictions = logistic_regression.predict(X_test)
lr_accuracy = accuracy_score(y_test, lr_predictions)
print("Logistic Regression accuracy:", lr_accuracy)

Logistic Regression accuracy: 0.78


In [ ]:
# Melatih dan menguji model SVM
svm.fit(X_train, y_train)
svm_predictions = svm.predict(X_test)
svm_accuracy = accuracy_score(y_test, svm_predictions)
print("SVM accuracy:", svm_accuracy)

SVM accuracy: 0.765


#Naive Bayes accuracy: 0.785
#Logistic Regression accuracy: 0.78
#SVM accuracy: 0.765
# Naive Bayes paling tinggi

#3.Menurut Anda apa saja yang perlu diubah/ditambah jika Anda diminta melakukan sentimen analisis menggunakan Bahasa Indonesia?

Sentimen analisis dalam Bahasa Indonesia:
Untuk melakukan sentimen analisis dalam Bahasa Indonesia, ada beberapa hal yang perlu dilakukan:
a. Memastikan dataset yang digunakan adalah dataset dalam Bahasa Indonesia yang relevan.

b. Menggunakan library dan modul yang mendukung pemrosesan teks Bahasa Indonesia, seperti NLTK dengan Indonesian Stemmer Factory.

c. Melakukan langkah-langkah preprocessing yang sesuai untuk Bahasa Indonesia, seperti tokenisasi kata, stemming, penghapusan stopwords, dan pembersihan teks yang spesifik untuk Bahasa Indonesia.

d. Memilih model klasifikasi yang tepat untuk sentimen analisis dalam Bahasa Indonesia, seperti Naive Bayes, SVM, atau model deep learning yang telah dilatih dengan data Bahasa Indonesia.

e. Melakukan pelabelan sentimen pada dataset Bahasa Indonesia yang digunakan, di mana sentimen yang diinginkan diwakili oleh label kelas yang sesuai (misalnya, "positif", "negatif", "netral").

f. Melatih dan menguji model klasifikasi menggunakan metode Bag of Words atau pendekatan lain yang sesuai untuk Bahasa Indonesia.